File: python/LC_904_FruitIntoBaskets.py
# https://leetcode.com/problems/fruit-into-baskets/
from typing import List


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        map = {}
        start, maxTotal = 0, 0
        for end in range(len(fruits)):
            map[fruits[end]] = map.get(fruits[end], 0)+1
            while len(map) > 2:
                map[fruits[start]] = map[fruits[start]] - 1
                if map[fruits[start]] == 0:
                    del map[fruits[start]]
                start += 1
            maxTotal = max(maxTotal, end-start+1)
        return maxTotal


sol = Solution()
print(sol.totalFruit([1, 2, 1]))
print(sol.totalFruit([0, 1, 2, 2]))
print(sol.totalFruit([1, 2, 3, 2, 2]))
print(sol.totalFruit([4, 1, 1, 1, 3, 1, 7, 5]))

File: java/src/sliding_window/LC_904_FruitIntoBaskets.java
package sliding_window;

import java.util.HashMap;

// https://leetcode.com/problems/fruit-into-baskets/
public class LC_904_FruitIntoBaskets {
    public int totalFruit(int[] fruits) {
        HashMap<Integer, Integer> fruitMap = new HashMap<>();
        int start = 0, maxTotal = 0;
        for (int end = 0; end < fruits.length; end++) {
            fruitMap.put(fruits[end], fruitMap.getOrDefault(fruits[end], 0) + 1);
            while (fruitMap.size() > 2) {
                int f = fruits[start++];
                fruitMap.put(f, fruitMap.get(f) - 1);
                if (fruitMap.get(f) == 0)
                    fruitMap.remove(f);
            }
            maxTotal = Math.max(maxTotal, end - start + 1);
        }
        return maxTotal;
    }
}
